Add masked_grad kw argument to to_dense

pearu opened the pull request:
As in the title.

The `masked_grad` kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked_grad=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked_grad==True` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked_grad=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked_grad=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked_grad=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96095



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

pearu committed:
Add masked kw argument to to_dense

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6702,19 +6702,19 @@
   variants: function
 
-- func: to_dense(Tensor self, ScalarType? dtype=None) -> Tensor
+- func: to_dense(Tensor self, ScalarType? dtype=None, bool? masked=False) -> Tensor
   variants: method
```
So, we are fine with making bc-breaking changes right now?

pearu replied:
In fact, the previous behavior for `to_dense` backward to ignore the unspecified elements (the `masked==True` case) was incorrect for all `torch` operations (`torch.sum`, `torch.mm`, etc) that support sparse returns.

nikitaved replied:
OK, I could clean up some gradchecks then once `sparse_mask` is differentiable.

pearu replied:
UPDATE: the default `masked` is `True` for BC as agreed in the 3/7/2023 meeting.

nikitaved commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -498,25 +498,48 @@ std::vector<Tensor> _to_cpu(TensorList tensors) {
+      if (masked.value_or(false)) {
+        TORCH_WARN("strided tensor to dense backward ignores masked=true");
+      }
+      // TODO: return grad as it is
       return grad.to_dense();
```
We should probably pass `masked` to `grad.to_dense()` as well to make higher order divs correct. In fact, what the lack of `gradgradcheck`s really bothers me with sparse...

pearu replied:
Fixed.

Re `gradgradcheck`: it should support sparse inputs similar to gradcheck. Please create an issue if you find that it is not the case.

nikitaved replied:
It should, but I've never seen explicit `gradgracheck` calls in the tests...

nikitaved commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -498,25 +498,48 @@ std::vector<Tensor> _to_cpu(TensorList tensors) {
       // TODO: add efficient BSR/BSC support for sparse_mask
-      const auto blocksize = at::DimVector(input_.values().sizes().slice(1, 2));
-      return grad.sparse_mask(input_.to_sparse()).to_sparse(input_layout, blocksize);
+      //const auto blocksize = at::DimVector(input_.values().sizes().slice(1, 2));
```
nit: to be removed?

pearu replied:
Fixed.

pearu committed:
Update on "Add masked kw argument to to_dense"

As in the title.

The masked kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked==False`.

The PR is BC-breaking in the sense that the masked semantics has been the default semantics for `to_dense` (its backward ignores unspecified elements in the input) and this PR enables the non-masked semantics as the default.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked=True)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked=True))
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Add masked kw argument to to_dense"

As in the title.

The masked kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked==False`.

The PR is BC-breaking in the sense that the masked semantics has been the default semantics for `to_dense` (its backward ignores unspecified elements in the input) and this PR enables the non-masked semantics as the default.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked=True)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked=True))
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

nikitaved commented on tools/autograd/derivatives.yaml:
```diff
@@ -1676,10 +1676,10 @@
 #
-- name: _to_dense(Tensor self, ScalarType? dtype=None) -> Tensor
-  self: to_dense_backward(grad, self)
+- name: _to_dense(Tensor self, ScalarType? dtype=None, bool? masked=False) -> Tensor
```
Does it have to be `optional<bool>` and not just `bool`?

albanD replied:
I think the only reason for keeping the optional is if you want to be able to differentiate when the user passed in a value or nothing (for deprecation purposes for example).
But if you don't, I agree that the optional is not needed.

nikitaved replied:
The default value is `False` though, maybe `nullopt` could be better, to exactly differentiate when the user provided the value and probably warn about bc-breaking changes?

pearu replied:
Using `nullopt` now and the default is now `True` for BC.

albanD commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -498,25 +498,47 @@ std::vector<Tensor> _to_cpu(TensorList tensors) {
+        TORCH_WARN("strided tensor to dense backward ignores masked=true");
+      }
+      // TODO: return grad as it is
+      return grad.to_dense(c10::nullopt, masked);
```
The autograd engine will do it for you if you don't IIRC ;)

pearu replied:
Now passing `dtype` to `to_dense` but there is no apparent effect for the given example.

pearu committed:
Update on "Add masked kw argument to to_dense"

As in the title.

The masked kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked==False`.

The PR is BC-breaking in the sense that the masked semantics has been the default semantics for `to_dense` (its backward ignores unspecified elements in the input) and this PR enables the non-masked semantics as the default.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked=True)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked=True))
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Add masked kw argument to to_dense"

As in the title.

The masked kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked==False`.

The PR is BC-breaking in the sense that the masked semantics has been the default semantics for `to_dense` (its backward ignores unspecified elements in the input) and this PR enables the non-masked semantics as the default.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked=True)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked=True))
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Add masked kw argument to to_dense"

As in the title.

The masked kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked==False` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

~The PR is BC-breaking in the sense that the masked semantics has been the default semantics for `to_dense` (its backward ignores unspecified elements in the input) and this PR enables the non-masked semantics as the default.~

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Add masked kw argument to to_dense"

As in the title.

The masked kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked==False` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

~The PR is BC-breaking in the sense that the masked semantics has been the default semantics for `to_dense` (its backward ignores unspecified elements in the input) and this PR enables the non-masked semantics as the default.~

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

amjames commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6711,16 +6711,20 @@
 - func: to_dense(Tensor self, ScalarType? dtype=None) -> Tensor
   variants: method
 
+# In to_dense.masked, the masked argument appears before dtype to avoid ambiguity with to_dense
```
Is there precedent for adding these types of overloads? Are they for the benefit of the C++ API  where named arguments do not exist?

cpuhrsch replied:
We don't really need an overload here. We could just add a keyword-only argument with default value `True` to the existing signature.

pearu replied:
@cpuhrsch, my initial approach used the keyword-only argument approach and it did work out for all tests except for test/jit/test_freezing.py, see https://pytorch.slack.com/archives/CFZ7V4DNZ/p1678140643591249 .

pearu replied:
Not using overloads anymore, so resolving.

amjames reviewed (commented):
The kwarg should really be called `masked_grad` or `masked_graident` no? 

It does nothing to the behavior of the method itself only affecting the backward behaviors so I think that should be clear from the name.

cpuhrsch commented:
On the name of the kwarg:
There is [sparse_grad for gather](https://pytorch.org/docs/master/generated/torch.gather.html?highlight=sparse_grad) and [sparse for Embedding](https://pytorch.org/docs/master/generated/torch.nn.Embedding.html?highlight=embedding#torch.nn.Embedding). None of these are great.

I agree that we should make it clear that this is influencing the gradient. I guess `masked_grad` is the closest to what we currently have?

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6711,16 +6711,20 @@
   variants: method
 
+# In to_dense.masked, the masked argument appears before dtype to avoid ambiguity with to_dense
+- func: to_dense.masked(Tensor self, bool? masked, ScalarType? dtype=None) -> Tensor
```
@pearu - can we add this as a kwarg-only (separate with ",*,") at the end of the signature? Otherwise we'll run into [forward compatibility](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy) issues since masked can be used positionally.

pearu replied:
@cpuhrsch I'd like to but this leads to the issue reported in https://pytorch.slack.com/archives/CFZ7V4DNZ/p1678140643591249 . If there is a better fix, I'd be happy to apply it.

pearu replied:
Fixed via https://github.com/pytorch/pytorch/pull/96095#discussion_r1135544478

amjames commented:
> On the name of the kwarg: There is [sparse_grad for gather](https://pytorch.org/docs/master/generated/torch.gather.html?highlight=sparse_grad) and [sparse for Embedding](https://pytorch.org/docs/master/generated/torch.nn.Embedding.html?highlight=embedding#torch.nn.Embedding). None of these are great.

I would actually say we should open an issue to change the name of `sparse_grad` for gather. It flies directly in the face of how we are differentiating between masked and sparse. 

That aside, I am advocating for `masked_grad` here as I think it is the best combination of simplicity and correctness. It is indicating that `masked` semantics should be applied to the `grad`ient. 

I don't want to get this hung up on naming as we can always resolve clarity issues with documentation, but I think this is a pretty small change to apply here. Changing it down the line introduces issues with BW compatibility so it is not as simple as it might seem on the surface.

We are working hard to clearly delineate between sparse and masked and this is an easy win on that front IMO.

cpuhrsch commented:
@amjames - At this point I don't think we have the luxury of changing that name anymore, but it does seem useful to have a well succinct description of what these keyword arguments do in general. Since we currently have only 3 of those in PyTorch from what I can tell, let's add this to the roadmap so we can track it and revisit it when we're prioritizing our next release.

amjames commented:
@cpuhrsch I figured as much for the existing cases. 

Given that it is a small number right now I would rather they be the exception to the rule considering the naming here. However as we have identified this is a root cause for masked semantics confusion I don't know how many more cases will appear requiring a similar kwarg. I  still prefer a more correct naming for new instances moving forward (starting with this case).

pearu commented:
I think using `masked_grad` or similar makes sense. 

`masked_grad=True` currently means that the backward will return `grad.sparse_mask(input)`.

Let me switch to a general discussion on supporting masked semantics.

First, in the context of Masked Tensors, we have a masked tensor as a pair `(data, mask)` where `data` is a tensor of any layout and dtype, `mask` is a tensor of any layout (in principle, however, currently `data.layout == mask.layout` is required and `data.indices() == mask.indices()` is forced) and `dtype=bool`.
Now, `masked_tensor.to_dense()` is defined as `MaskedTensor(masked_tensor.get_data().to_dense(), masked_tensor.get_mask().to_dense())`. IIUC, the backward of `MaskedTensor.to_dense` (currently not implemented) will return `MaskedTensor(grad, input.get_mask())`, that is, the `masked_grad` kw argument in `to_dense` becomes redundant for Masked Tensors.

Alternatively, the boolean kw argument could be generalized to a `mask` tensor argument so that we would have `torch.Tensor.to_dense(input, mask=None)` and the `to_dense` backward implementation would simply return `grad.sparse_mask(mask)` that is applicable for any input and grad layout (here `sparse_mask(None)` would be a no-op as `mask=None` means `mask=ones_like(input)`). I realize that this alternative may be off the chosen Masked Tensor path, so one can consider this as "just for the record" idea, but it also provides a more general framework in eliminating `masked_grad` kw (`to_dense(input, masked_grad=True)` would be equivalent to `to_dense(input, mask=input)`) as well as in considering ordinary tensors as special cases of masked tensors (`to_dense(input, masked_grad=False)` would be equivalent to `to_dense(input, mask=ones_like(input))`).

In sum, the `masked_grad` kw argument in `to_dense` (or `gather`, etc) ought to be a temporary need (and it could be removed once Masked Tensors autograd support is completed and if there would be no BC requirements which we do...). However, internally, Masked Tensors support could find it useful, for instance, `MaskedTensor.to_dense` can be defined as `MaskedTensor(masked_tensor.get_data().to_dense(mask=masked_tensor.get_mask()), masked_tensor.get_mask().to_dense())` (notice that the boolean kw argument `masked_grad` would be insufficient here). The same pattern would apply to other operations as well: `torch.unary_op(input, mask=None)` implies that the backward of `unary_op` will return `grad.sparse_mask(mask)` (with `mask=None` making `sparse_mask` a no-op). Just a thought...

pearu committed:
Update on "Add masked kw argument to to_dense"

As in the title.

The masked kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked==False` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

~The PR is BC-breaking in the sense that the masked semantics has been the default semantics for `to_dense` (its backward ignores unspecified elements in the input) and this PR enables the non-masked semantics as the default.~

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Add masked_grad kw argument to to_dense"

As in the title.

The `masked_grad` kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked_grad=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked_grad==True` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked_grad=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked_grad=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked_grad=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -737,11 +737,11 @@ void ComputeSubgraphInMKLDNN(Node* subgraph_node) {
-            ->create(
-                c10::Symbol::fromQualString("aten::to_dense"), {v, none_value})
-            ->insertAfter(subgraph_node);
+    auto from_mkldnn = graph
+                           ->create(
+                               c10::Symbol::fromQualString("aten::to_dense"),
+                               {v, none_value, none_value})
+                           ->insertAfter(subgraph_node);
```
This changeset fixes the following failure:
```
2023-03-06T20:19:12.8348868Z RuntimeError: 0 INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/torch/csrc/jit/ir/alias_analysis.cpp":621, please report a bug to PyTorch. We don't have an op for aten::to_dense but it isn't a special case.  Argument types: Tensor, NoneType,
2023-03-06T20:19:12.8348877Z 
2023-03-06T20:19:12.8348950Z Candidates:
2023-03-06T20:19:12.8349167Z 	aten::to_dense(Tensor self, ScalarType? dtype=None, bool? masked=False) -> Tensor
```

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6709,19 +6709,19 @@
   variants: function
 
-- func: to_dense(Tensor self, ScalarType? dtype=None) -> Tensor
+- func: to_dense(Tensor self, *, ScalarType? dtype=None, bool? masked_grad=None) -> Tensor
```
Ah, you can't make dtype kwarg-only. That'll be BC breaking technically and might cause all kinds of headaches. But I think it's ok to make masked_grad kwarg-only.

pearu replied:
Right. Fixed.

pearu committed:
Update on "Add masked_grad kw argument to to_dense"

As in the title.

The `masked_grad` kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked_grad=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked_grad==True` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked_grad=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked_grad=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked_grad=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Add masked_grad kw argument to to_dense"

As in the title.

The `masked_grad` kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked_grad=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked_grad==True` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked_grad=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked_grad=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked_grad=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

cpuhrsch commented on torch/_tensor_docs.py:
```diff
@@ -5360,10 +5360,16 @@ def callable(a, b) -> number
     "to_dense",
     r"""
-to_dense() -> Tensor
+to_dense(dtype=None, *, masked_grad=None) -> Tensor
```
nit: Did you mean `masked_grad=True`?

pearu replied:
I guess in the docs it is ok to use `masked_grad=True` (I'll fix it). Otherwise, in the implementations, we use `masked_grad=None` to make the deprecation and/or changing of the default processes easier.

pearu replied:
Fixed.

pearu committed:
Update on "Add masked_grad kw argument to to_dense"

As in the title.

The `masked_grad` kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked_grad=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked_grad==True` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked_grad=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked_grad=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked_grad=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

cpuhrsch commented:
Some of these changes worry me w.r.t. TorchScript and FC/BW compatability. Since we're adding a new argument functions like mkldnn_to_dense or _to_dense I'm worried that somewhere a version with less arguments is called and baked in. @pearu can you resubmit this as a regular PR (not using ghstack) so I can try using the usual land workflow and get internal CI signal first? I think we should avoid the usual GH workflow before getting additional signal.

nikitaved commented:
@cpuhrsch , I think you can still import ghstack changes:

pearu commented:
@cpuhrsch here is a regular copy of this PR: https://github.com/pytorch/pytorch/pull/96810

nikitaved commented on test/test_sparse.py:
```diff
@@ -561,7 +557,7 @@ def test_tensor(x, res):
-                return x.to_dense()
+                return x.to_dense(masked_grad=gradcheck.masked)
             x.requires_grad_(True)
             gradcheck(fn, (x,), check_sparse_nnz=True)
```
This tests both `masked=True` and `masked=False`, right?

pearu replied:
Yes, see the `gradcheck_semantics` decorator.

nikitaved commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -526,21 +555,21 @@ Tensor to_dense_backward(const Tensor& grad, const Tensor& input_) {
 
 Tensor to_mkldnn_backward(const Tensor& grad, const Tensor& input_) {
   AT_ASSERT(input_.layout() == c10::kStrided);
-  return grad.to_dense(input_.scalar_type());
+  return grad.to_dense();
```
Why was this line modified?

pearu replied:
It is a typo. I'll restore it.

pearu replied:
Fixed.

nikitaved commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -1756,7 +1786,7 @@ Tensor sparse_compressed_to_sparse(const Tensor& self, c10::optional<c10::Layout
   }
   switch (layout_) {
   case kStrided:
-    return sparse_compressed_to_dense(self);
+    return sparse_compressed_to_dense(self, /*dtype=*/c10::nullopt, /*masked_grad=*/c10::nullopt);
```
Wouldn't it be more clear to pass explicit `true/false` instead? All while leaving `nullopt` for something to indicate that the user (not the dev) did not provide any value.

pearu replied:
No. I am using `nullopt` to ensure the default behavior. In addition, `sparse_compressed_to_dense` does not use `masked_grad` value anyway.

Hmm, it is not clear to me how the specification of `dtype` was not required before...

pearu replied:
Actually, the right thing to do here would be to add `masked_grad` kw argument to `to_sparse` method as well so that the correct value could be propagated to `to_dense` call when the specified layout is `torch.strided`. However, changing `to_sparse` API is OT here and I suggest we leave the code as it is for `masked_grad` visibility here.

pearu committed:
Update on "Add masked_grad kw argument to to_dense"

As in the title.

The `masked_grad` kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked_grad=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked_grad==True` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked_grad=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked_grad=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked_grad=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu committed:
Update on "Add masked_grad kw argument to to_dense"

As in the title.

The `masked_grad` kw argument is required for `to_dense` backward to distinguish the expected semantics of sparse tensors. `masked_grad=True` means that the `to_dense` backward will apply a mask to the returned gradient where the mask is defined by the input indices. The default semantics implies `masked_grad==True` for BC but see the [comment](https://github.com/pytorch/pytorch/pull/96095/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR501-R513) in `to_dense_backward`.

As a consequence, existing code that is run through autograd engine must replace `.to_dense()` calls with `.to_dense(masked_grad=False)`. For example,
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense())
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense())
```
(recall, gradcheck has `masked=False` as default) must be updated to
```python
torch.autograd.gradcheck(lambda x: torch.sum(x, [0]).to_dense(masked_grad=False))
torch.autograd.gradcheck(lambda x: torch.sparse.sum(x, [0]).to_dense(masked_grad=True), masked=True)
```

Fixes https://github.com/pytorch/pytorch/issues/95550




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7 jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

pearu commented:
@pytorchbot merge
